crates/roc_std/src/roc_str.rs

```diff
@@ -697,19 +793,216 @@ impl From<SendSafeRocStr> for RocStr {
+    }
+
+    fn ptr_to_allocation(&self) -> *mut usize {
+        // these are the same because the alignment of u8 is just 1
+        self.ptr_to_refcount()
+    }
+
+    fn ptr_to_refcount(&self) -> *mut usize {
+        if self.is_seamless_slice() {
+            unsafe { ((self.capacity_or_ref_ptr << 1) as *mut usize).sub(1) }
+        } else {
+            unsafe { self.ptr_to_first_elem().cast::<usize>().sub(1) }
+        }
+    }
+
+    fn as_bytes(&self) -> &[u8] {
+        unsafe { std::slice::from_raw_parts(self.ptr_to_first_elem(), self.len()) }
+    }
+
+    fn as_str(&self) -> &str {
+        unsafe { std::str::from_utf8_unchecked(self.as_bytes()) }
+    }
+
+    fn is_unique(&self) -> bool {
+        if self.capacity() == 0 {
+            return false;
+        }
+
+        let ptr = self.ptr_to_refcount();
+        let rc = unsafe { std::ptr::read(ptr) as isize };
+
+        rc == isize::MIN
+    }
+
+    fn is_readonly(&self) -> bool {
+        if self.capacity() == 0 {
+            return true;
+        }
+
+        let ptr = self.ptr_to_refcount();
+        let rc = unsafe { std::ptr::read(ptr) as isize };
+
+        rc == 0
+    }
+
+    fn set_readonly(&mut self) {
+        assert_ne!(self.capacity(), 0);
+
+        let ptr = self.ptr_to_refcount();
+        unsafe { std::ptr::write(ptr, 0) }
+    }
+
+    fn inc(&mut self, n: usize) {
+        let ptr = self.ptr_to_refcount();
+        unsafe {
+            let value = std::ptr::read(ptr);
+            std::ptr::write(ptr, Ord::max(0, ((value as isize) + n as isize) as usize));
+        }
+    }
+
+    fn dec(&mut self) {
+        if self.capacity() == 0 {
+            // no valid allocation, elements pointer is dangling
+            return;
+        }
+
+        let ptr = self.ptr_to_refcount();
+        unsafe {
+            let value = std::ptr::read(ptr) as isize;
+            match value {
+                0 => {
+                    // static lifetime, do nothing
+                }
+                isize::MIN => {
+                    // refcount becomes zero; free allocation
+                    crate::roc_dealloc(self.ptr_to_allocation().cast(), 1);
+                }
+                _ => {
+                    std::ptr::write(ptr, (value - 1) as usize);
+                }
+            }
+        }
+    }
+
+    fn with_capacity(cap: usize) -> Self {
+        let mut this = Self {
+            elements: NonNull::dangling(),
+            length: 0,
+            capacity_or_ref_ptr: 0,
+        };
+
+        this.reserve(cap);
+
+        this
+    }
+
+    /// Increase a BigString's capacity by at least the requested number of elements (possibly more).
+    ///
+    /// May return a new BigString, if the provided one was not unique.
+    fn reserve(&mut self, n: usize) {
+        let align = std::mem::size_of::<usize>();
+        let desired_cap = self.len() + n;
+        let desired_alloc = align + desired_cap;
+
+        if self.is_unique() {
```

Comment:
```suggestion
        if self.is_unique() && !self.is_seamless_slice() {
```

Seamless slice I think the realloc will be wrong. The bytes of the slice aren't at the start of the allocation.

Same with the list code I'm sure.